Author: Ben Hutchings <ben@decadent.org.uk>
Description: Add test case for direct mode and extra output files
Bug-Debian: https://bugs.debian.org/853834

--- a/test.sh
+++ b/test.sh
@@ -2231,6 +2231,27 @@ EOF
     if [ -n "$data" ]; then
         test_failed "$manifest contained ignored header: $data"
     fi
+
+    # -------------------------------------------------------------------------
+    TEST "Identical source and object files, different dependencies"
+
+    mkdir -p a/include b/include
+    echo '#include "header.h"' > a/source.c
+    backdate a/include/header.h
+    ln a/source.c b/source.c
+    ln a/include/header.h b/include/header.h
+
+    (cd a && $CCACHE_COMPILE -MD -I$PWD/include -c source.c)
+    (cd b && $CCACHE_COMPILE -MD -I$PWD/include -c source.c)
+    expect_equal_object_files a/source.o b/source.o
+    if cmp -s a/source.d b/source.d; then
+	test_failed "Dependencies unexpectedly identical"
+    fi
+    (cd a && $CCACHE_COMPILE -MD -I$PWD/include -c source.c)
+    expect_equal_object_files a/source.o b/source.o
+    if cmp -s a/source.d b/source.d; then
+	test_failed "Wrong dependencies restored from cache"
+    fi
 }
 
 # =============================================================================
